Kernel/Bus/USB/USBConfiguration.cpp

```diff
@@ -15,61 +16,95 @@ namespace Kernel::USB {
-                dbgln("Endpoint Descriptor {}", endpoint);
-                dbgln("Endpoint Address: {}", endpoint_descriptor->endpoint_address);
-                dbgln("Endpoint Attribute Bitmap: {:08b}", endpoint_descriptor->endpoint_attributes_bitmap);
-                dbgln("Endpoint Maximum Packet Size: {}", endpoint_descriptor->max_packet_size);
-                dbgln("Endpoint Poll Interval (in frames): {}", endpoint_descriptor->poll_interval_in_frames);
+                dbgln("Interface Descriptor:");
+                dbgln("  interface_id: {:02x}", interface_descriptor.interface_id);
+                dbgln("  alternate_setting: {:02x}", interface_descriptor.alternate_setting);
+                dbgln("  number_of_endpoints: {:02x}", interface_descriptor.number_of_endpoints);
+                dbgln("  interface_class_code: {:02x}", interface_descriptor.interface_class_code);
+                dbgln("  interface_sub_class_code: {:02x}", interface_descriptor.interface_sub_class_code);
+                dbgln("  interface_protocol: {:02x}", interface_descriptor.interface_protocol);
+                dbgln("  interface_string_descriptor_index: {}", interface_descriptor.interface_string_descriptor_index);
             }
 
-            endpoint_descriptors.unchecked_append(*endpoint_descriptor);
+            TRY(m_interfaces.try_empend(*this, interface_descriptor));
+            current_interface = &m_interfaces.last();
+
+            break;
         }
 
-        USBInterface device_interface(*this, *interface_descriptor, move(endpoint_descriptors));
-        m_interfaces.unchecked_append(move(device_interface));
-        interface_descriptor = reinterpret_cast<USBInterfaceDescriptor*>(reinterpret_cast<u8*>(interface_descriptor) + sizeof(USBInterfaceDescriptor) + interface_descriptor->number_of_endpoints * sizeof(USBEndpointDescriptor));
+        case DESCRIPTOR_TYPE_ENDPOINT: {
+            auto endpoint_descriptor = TRY(read_descriptor.operator()<USBEndpointDescriptor>(descriptor_header));
+
+            if constexpr (USB_DEBUG) {
+                dbgln("Endpoint Descriptor:");
+                dbgln("  Endpoint Address: {}", endpoint_descriptor.endpoint_address);
+                dbgln("  Endpoint Attribute Bitmap: {:08b}", endpoint_descriptor.endpoint_attributes_bitmap);
+                dbgln("  Endpoint Maximum Packet Size: {}", endpoint_descriptor.max_packet_size);
+                dbgln("  Endpoint Poll Interval (in frames): {}", endpoint_descriptor.poll_interval_in_frames);
+            }
+
+            if (current_interface == nullptr)
+                return EINVAL;
+
+            TRY(current_interface->add_endpoint_descriptor({}, endpoint_descriptor));
+
+            break;
+        }
+
+        default:
+            dbgln_if(USB_DEBUG, "Skipping descriptor of unknown type {}", descriptor_header.descriptor_type);
+            TRY(stream.seek(descriptor_header.length, SeekMode::FromCurrentPosition));
+            break;
+        }
```

Comment:
Should we compare number_of_interfaces with m_interfaces.size() here?

Reply:
That's probably a good idea, will add that in a follow-up.

Reply:
Ah no, we shouldn't compare this as alternate settings exist. Alternate settings are like USB configurations for single interfaces so you can choose what interface you get. These alternative settings share their interface number and have bAlternateSetting set to their alternate setting number.

Reply:
We currently don't handle alternative settings at all, so that is also something that needs to be fixed.